{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'biocodegff'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-3d528e4fd152>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mbiothings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mbiocodegff\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mbiocodeutils\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'biocodegff'"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    " \n",
    "\"\"\"\n",
    "This is a script to convert GenBank flat files to GFF3 format with a specific focus on\n",
    "maintaining as much annotation information as possible.\n",
    " \n",
    "This is not guaranteed to convert all features, but warnings will be printed wherever possible\n",
    "for features which aren't included.\n",
    " \n",
    "This is written to handle multi-entry GBK files\n",
    " \n",
    "Caveats:\n",
    "- Because the GBK flatfile format doesn't explicitly model parent/child features, this script\n",
    "  links them using the expected format convention of shared /locus_tag entries for each feature\n",
    "  of the gene graph (gene, mRNA, CDS)\n",
    " \n",
    "Author: Joshua Orvis (jorvis AT gmail)\n",
    "\"\"\"\n",
    " \n",
    "import argparse\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "from Bio import SeqIO\n",
    "import biothings\n",
    "import biocodegff\n",
    "import biocodeutils\n",
    " \n",
    "def main():\n",
    "    parser = argparse.ArgumentParser( description='Convert GenBank flat files to GFF3 format')\n",
    " \n",
    "    ## output file to be written\n",
    "    parser.add_argument('-i', '--input_file', type=str, required=True, help='Path to an input GBK file' )\n",
    "    parser.add_argument('-o', '--output_file', type=str, required=False, help='Path to an output GFF file to be created' )\n",
    "    parser.add_argument('--with_fasta', dest='fasta', action='store_true', help='Include the FASTA section with genomic sequence at end of file.  (default)' )\n",
    "    parser.add_argument('--no_fasta', dest='fasta', action='store_false' )\n",
    "    parser.set_defaults(fasta=True)\n",
    "    args = parser.parse_args()\n",
    " \n",
    "    ## output will either be a file or STDOUT\n",
    "    ofh = sys.stdout\n",
    "    if args.output_file is not None:\n",
    "        ofh = open(args.output_file, 'wt')\n",
    " \n",
    "    ofh.write(\"##gff-version 3\\n\")\n",
    " \n",
    "    assemblies = dict()\n",
    "    current_assembly = None\n",
    "    current_gene = None\n",
    "    current_RNA = None\n",
    " \n",
    "    rna_count_by_gene = defaultdict(int)\n",
    "    exon_count_by_RNA = defaultdict(int)\n",
    " \n",
    "    seqs_pending_writes = False\n",
    " \n",
    "    features_skipped_count = 0\n",
    " \n",
    "    # each gb_record is a SeqRecord object\n",
    "    for gb_record in SeqIO.parse(open(\"/mnt/students/spiders/annotations/spiders/GCA_001188405.1_Lrec_1.0_genomic.gbff\", \"r\"), \"genbank\"):\n",
    "        mol_id = gb_record.name\n",
    " \n",
    "        if mol_id not in assemblies:\n",
    "            assemblies[mol_id] = biothings.Assembly( id=mol_id )\n",
    " \n",
    "        if len(str(gb_record.seq)) > 0:\n",
    "            seqs_pending_writes = True\n",
    "            assemblies[mol_id].residues = str(gb_record.seq)\n",
    "            assemblies[mol_id].length = len(str(gb_record.seq))\n",
    " \n",
    "        current_assembly = assemblies[mol_id]\n",
    "             \n",
    "        # each feat is a SeqFeature object\n",
    "        for feat in gb_record.features:\n",
    "            #print(feat)\n",
    "            fmin = int(feat.location.start)\n",
    "            fmax = int(feat.location.end)\n",
    " \n",
    "            if feat.location.strand == 1:\n",
    "                strand = '+'\n",
    "            elif feat.location.strand == -1:\n",
    "                strand = '-'\n",
    "            else:\n",
    "                raise Exception(\"ERROR: unstranded feature encountered: {0}\".format(feat))\n",
    " \n",
    "            #print(\"{0} located at {1}-{2} strand:{3}\".format( locus_tag, fmin, fmax, strand ) )\n",
    "            if feat.type == 'source':\n",
    "                continue\n",
    "             \n",
    "            if feat.type == 'gene':\n",
    "                # print the previous gene (if there is one)\n",
    "                if current_gene is not None:\n",
    "                    gene.print_as(fh=ofh, source='GenBank', format='gff3')\n",
    "                 \n",
    "                locus_tag = feat.qualifiers['locus_tag'][0]\n",
    "                gene = biothings.Gene( id=locus_tag )\n",
    "                gene.locate_on( target=current_assembly, fmin=fmin, fmax=fmax, strand=strand )\n",
    "                current_gene = gene\n",
    "                current_RNA = None\n",
    " \n",
    "            elif feat.type == 'mRNA':\n",
    "                locus_tag = feat.qualifiers['locus_tag'][0]\n",
    "                rna_count_by_gene[locus_tag] += 1\n",
    "                feat_id = \"{0}.mRNA.{1}\".format( locus_tag, rna_count_by_gene[locus_tag] )\n",
    "                 \n",
    "                mRNA = biothings.mRNA( id=feat_id, parent=current_gene )\n",
    "                mRNA.locate_on( target=current_assembly, fmin=fmin, fmax=fmax, strand=strand )\n",
    "                gene.add_mRNA(mRNA)\n",
    "                current_RNA = mRNA\n",
    " \n",
    "                if feat_id in exon_count_by_RNA:\n",
    "                    raise Exception( \"ERROR: two different RNAs found with same ID: {0}\".format(feat_id) )\n",
    "                else:\n",
    "                    exon_count_by_RNA[feat_id] = 0\n",
    " \n",
    "            elif feat.type == 'tRNA':\n",
    "                locus_tag = feat.qualifiers['locus_tag'][0]\n",
    "                rna_count_by_gene[locus_tag] += 1\n",
    "                feat_id = \"{0}.tRNA.{1}\".format( locus_tag, rna_count_by_gene[locus_tag] )\n",
    "                 \n",
    "                tRNA = biothings.tRNA( id=feat_id, parent=current_gene )\n",
    "                tRNA.locate_on( target=current_assembly, fmin=fmin, fmax=fmax, strand=strand )\n",
    "                gene.add_tRNA(tRNA)\n",
    "                current_RNA = tRNA\n",
    " \n",
    "                if feat_id in exon_count_by_RNA:\n",
    "                    raise Exception( \"ERROR: two different RNAs found with same ID: {0}\".format(feat_id) )\n",
    "                else:\n",
    "                    exon_count_by_RNA[feat_id] = 0\n",
    " \n",
    "            elif feat.type == 'rRNA':\n",
    "                locus_tag = feat.qualifiers['locus_tag'][0]\n",
    "                rna_count_by_gene[locus_tag] += 1\n",
    "                feat_id = \"{0}.rRNA.{1}\".format( locus_tag, rna_count_by_gene[locus_tag] )\n",
    "                 \n",
    "                rRNA = biothings.rRNA( id=feat_id, parent=current_gene )\n",
    "                rRNA.locate_on( target=current_assembly, fmin=fmin, fmax=fmax, strand=strand )\n",
    "                gene.add_rRNA(rRNA)\n",
    "                current_RNA = rRNA\n",
    " \n",
    "                if feat_id in exon_count_by_RNA:\n",
    "                    raise Exception( \"ERROR: two different RNAs found with same ID: {0}\".format(feat_id) )\n",
    "                else:\n",
    "                    exon_count_by_RNA[feat_id] = 0\n",
    "             \n",
    "            elif feat.type == 'CDS':\n",
    "                locus_tag = feat.qualifiers['locus_tag'][0]\n",
    "                # If processing a prokaryotic GBK, we'll encounter CDS before mRNA, so we have to\n",
    "                #  manually make one\n",
    "                if current_RNA is None:\n",
    "                    feat_id = \"{0}.mRNA.{1}\".format( locus_tag, rna_count_by_gene[locus_tag] )\n",
    "                    mRNA = biothings.mRNA( id=feat_id, parent=current_gene )\n",
    "                    mRNA.locate_on( target=current_assembly, fmin=fmin, fmax=fmax, strand=strand )\n",
    "                    gene.add_mRNA(mRNA)\n",
    "                    current_RNA = mRNA\n",
    "                 \n",
    "                exon_count_by_RNA[current_RNA.id] += 1\n",
    "                cds_id = \"{0}.CDS.{1}\".format( current_RNA.id, exon_count_by_RNA[current_RNA.id] )\n",
    "                current_CDS_phase = 0\n",
    "                 \n",
    "                for loc in feat.location.parts:\n",
    "                    subfmin = int(loc.start)\n",
    "                    subfmax = int(loc.end)\n",
    "                     \n",
    "                    CDS = biothings.CDS( id=cds_id, parent=current_RNA )\n",
    "                    CDS.locate_on( target=current_assembly, fmin=subfmin, fmax=subfmax, strand=strand, phase=current_CDS_phase )\n",
    "                    current_RNA.add_CDS(CDS)\n",
    " \n",
    "                    # calculate the starting phase for the next CDS feature (in case there is one)\n",
    "                    # 0 + 6 = 0     TTGCAT\n",
    "                    # 0 + 7 = 2     TTGCATG\n",
    "                    # 1 + 6 = 1     TTGCAT\n",
    "                    # 2 + 7 = 1     TTGCATG\n",
    "                    # general: 3 - ((length - previous phase) % 3)\n",
    "                    current_CDS_phase = 3 - (((subfmax - subfmin) - current_CDS_phase) % 3)\n",
    "                    if current_CDS_phase == 3:\n",
    "                        current_CDS_phase = 0\n",
    " \n",
    "                    exon_id = \"{0}.exon.{1}\".format( current_RNA.id, exon_count_by_RNA[current_RNA.id] )\n",
    "                    exon = biothings.Exon( id=exon_id, parent=current_RNA )\n",
    "                    exon.locate_on( target=current_assembly, fmin=subfmin, fmax=subfmax, strand=strand )\n",
    "                    current_RNA.add_exon(exon)\n",
    "                    exon_count_by_RNA[current_RNA.id] += 1\n",
    "                 \n",
    "            else:\n",
    "                print(\"WARNING: The following feature was skipped:\\n{0}\".format(feat))\n",
    "                features_skipped_count += 1\n",
    " \n",
    "        # don't forget to do the last gene, if there were any\n",
    "        if current_gene is not None:\n",
    "            gene.print_as(fh=ofh, source='GenBank', format='gff3')\n",
    " \n",
    "    if args.fasta is True:\n",
    "        if seqs_pending_writes is True:\n",
    "            ofh.write(\"##FASTA\\n\")\n",
    "            for assembly_id in assemblies:\n",
    "                ofh.write(\">{0}\\n\".format(assembly_id))\n",
    "                ofh.write(\"{0}\\n\".format(biocodeutils.wrapped_fasta(assemblies[assembly_id].residues)))\n",
    " \n",
    "    if features_skipped_count > 0:\n",
    "        print(\"Warning: {0} unsupported feature types were skipped\".format(features_skipped_count))\n",
    " \n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
